### DISCUSSION TITLE
Find the most optimal DOSEMU2-compatible Windows display driver

### DISCUSSION THREAD

**Discussion by jschwartzenberg** [dosemu2/dosemu2#1075] [Ideas]
@stsp I hope you don't mind I create a ticket here to track issues with different Windows 3.1 applications and different display drivers. (I guess these kinds of issues are not emulation related.)

MS Office seems to have an issue with `t800-5.drv` from `tvgaw31b.zip`:
![afbeelding](https://user-images.githubusercontent.com/877112/77829175-a4714400-7120-11ea-92bd-6154bea6d975.png)

**Comment by stsp** [comment]
Why not to use #265 for that?

**Comment by jschwartzenberg** [comment]
Because that one will quite likely also require changes to DOSEMU2 itself. There are no generic VESA drivers that can do this, so it makes sense to support the Trident drivers available that can do this. It seems to be a matter of adding more memory to the emulated Trident chip (if that's not already there) and interpreting the mode that the higher bit depth drivers use correctly. When you use them now, something legible already shows up.

**Comment by stsp** [comment]
> There are no generic VESA drivers that can do this,

Wait, wasn't there one written by @Baron-von-Riedesel ?

**Reply by skitt** [reply]
> > There are no generic VESA drivers that can do this,
> 
> Wait, wasn't there one written by @Baron-von-Riedesel ?

Yes, that’s https://www.japheth.de/Download/Win16/svgaptch.zip which patches `SVGA256.DRV` to use VESA calls. It’s limited to the resolutions supported by `SVGA256.DRV`: 640×480, 800×600 and 1024×768 in 256 colours.

**Comment by jschwartzenberg** [comment]
With the Trident driver that is now installed through the install-win31 repository, Klik & Play (demo version from here: https://archive.org/details/SimCity_201708) is very slow. So that's a good test case.

**Reply by stsp** [reply]
And with VESA?

**Reply by jschwartzenberg** [reply]
I just tried these: https://archive.org/details/SVGA_EXE

Much faster!

**Reply by stsp** [reply]
So once again, why to care about
trident instead of using vesa driver?

**Reply by stsp** [reply]
VESA gives you an LFB, which should
be the fastest solution among anything
un-accellerated in hw.

**Reply by jschwartzenberg** [reply]
Note that I used the 800x600 with 256 colors driver for 512kB boards. The 640x480 Trident driver seems faster.
These sets are similar.
https://archive.org/details/TRIDNT93_ZIP
https://archive.org/download/Trident-TVGA-8900-Utility-Drivers-Disk

~~Another driver to try would be SDD.~~ While SDD contains a win31 utility (to manage its DOS stuff) it only has win9x drivers.

**Reply by jschwartzenberg** [reply]
> VESA gives you an LFB, which should
> be the fastest solution among anything
> un-accellerated in hw.

Ah I didn't know that! That explains :)
I just hadn't done any testing either so I thought it would at least be something to check.

**Comment by jschwartzenberg** [comment]
I think this discussion is closed then. For performance the SVGA driver is by far better. I'll replace the driver in `install-win31`.

**Reply by jschwartzenberg** [reply]
Only for higher bit depths, supporting a better Trident might make sense.

**Comment by jschwartzenberg** [comment]
Is it me or do these SVGA drivers lack the ability to run the "MS-DOS" prompt in a window? Or maybe it's something else that is needed for that?

**Comment by stsp** [comment]
Its not possible in 286 mode afaik.

**Reply by jschwartzenberg** [reply]
Yeah then I misremember that I ever saw that working in DOSEMU I guess. 386 mode doesn't support MS-DOS prompts yet I guess?

**Reply by stsp** [reply]
Likely never will.
SHELL vxd is undocumented, and
full 386 support of win31 is not a
priority.
Note that SimCity200/win installer
says 386 mode is needed for the
game, but under dosemu2 it works
in 286 mode beautifully. :)
In 386 mode it doesn't work correctly
because vtd timer is not implemented,
but in 286 mode it works fine.

**Reply by andrewbird** [reply]
If it does then I'd expect our MFS redirector and LFN support to have real problems. This [note](https://web.archive.org/web/20061001224249/http://www.coli.uni-saarland.de/~eric/stuff/soft/specials/win3.x-dosext-freedos-notes.txt) from Eric Auer was recently referenced in a post to freedos-devel touting Win3x support. From it I see that Win3x wants to have private instance data for each 'Dosbox', specifically things like CDS and SFT, so you can have different current directories and open files in each. Many calls to the redirector supply the CDS and SFT so presumably would be passed the instanced data and should be okay, but others have to discover it and as far as I know we can only find out where the global one is. I think our LFN is in an even more disadvantaged situation as it has no such help.

**Reply by stsp** [reply]
We never "cache" these structures.
And any way of "getting" them (either
via DOS API or from LoL) will return
the correct one.
However, nothing like this will happen
as this is likely not the case for 286 mode.

**Reply by andrewbird** [reply]
Our posts crossed, I was referring to `386 mode doesn't support MS-DOS prompts yet I guess?`

My reading of the note suggests in enhanced mode a private copy of CDS per 'Dosbox', so potentially if you started two you could 'cd \dos' in one and 'cd \tmp' in the other. If the redirector were to call int2f/1217, I suspect it would get the global CDS and contains current directory that was set before windows was started.

**Reply by stsp** [reply]
> I suspect it would get the global CDS and contains current directory that was set before windows was started.

I don't think so.
But under dosemu2 we'll not find out.

**Reply by jschwartzenberg** [reply]
> Note that SimCity200/win installer
> says 386 mode is needed for the
> game, but under dosemu2 it works
> in 286 mode beautifully. :)

Any idea where the suggestion comes from that 386 mode is needed and/or why dosemu2 does handle the game fine in 286 mode?

**Reply by stsp** [reply]
Installer says that 386 mode is
needed for the game if you run
it (installer) in 286 mode.
But the game itself works in 286
mode properly.

**Reply by jschwartzenberg** [reply]
Yep, I saw exactly that. But why would they put that note in the installer? Is there something peculiar about win31 in 286 mode that still allows the game to run or is that note just incorrect?

**Reply by stsp** [reply]
This is what I don't know.
But in 386 mode the lack of timer
vxd prevents it from working,
whereas in 286 mode all drivers
are in ring3 and work fine.
Also note that there is a slight
difference between standard and
286 mode. What dosemu calls
286 mode is actually the "standard"
mode. Try running `system\krnl286.exe`
for 286 mode and see if the game
still works.

**Reply by jschwartzenberg** [reply]
> Try running `system\krnl286.exe`
> for 286 mode and see if the game
> still works.

The game still runs fine then. Its installer however breaks with this error:
![Screenshot_20210815_150812](https://user-images.githubusercontent.com/877112/129479672-b8345477-5d1b-4d82-853f-7502d3834095.png)

Also this is printed:
```
ERROR: LDT: Unimplemented rep
```

**Comment by stsp** [comment]
We've seen that pagefault already, its
a kernel bug. Please wait until all kernels
are updated.

**Reply by jschwartzenberg** [reply]
Ah even with `$_cpu_vm_dpmi="native"`? I didn't have problems in that mode before.

**Comment by stsp** [comment]
No, something else then.
Worth a new ticket.

**Reply by jschwartzenberg** [reply]
The installer runs fine through a simple `win` though. Now I'm curious what's the difference between `win /2` and `system\krnl286` :)

**Comment by stsp** [comment]
No one (besides ms) knows the
technical difference, but krnl386
implements both std and enh modes
while krnl286 implements std mode
for 286 cpu. Maybe it avoids the
386-specific segment attributes, but
we should just treat it as yet another
kernel for testing (besides winos2 one).

**Reply by jschwartzenberg** [reply]
Alright :)
I'm curious if today there's still a reason to install the winos2 kernel. I recently checked it out. Possibly these kernels could installed side by side even? :)

**Comment by jschwartzenberg** [comment]
It seems Sim Ant might not like the SVGA driver:
![Screenshot_20210815_162453](https://user-images.githubusercontent.com/877112/129481717-8015d1de-b574-444b-b267-fb8f646196b4.png)

**Comment by stsp** [comment]
So I tried the SVGA driver from here:
https://winworldpc.com/download/48c2ac3c-4918-c39a-11c3-a4e284a2c3a5
which presumably have no Japheth
patch. But all modes work fine.
Confused.
Can someone please re-check?

**Reply by jschwartzenberg** [reply]
If I remember correctly, that patch is only needed for pass-through mode with VESA hardware. That driver also lists explicit support for that Trident chip, so I guess it switches the mode that way with vgaemu.

**Reply by stsp** [reply]
> If I remember correctly, that patch is only needed for pass-through mode with VESA hardware.

Ah, makes sense.
So with vgaemu, no patch is needed right?

**Reply by jschwartzenberg** [reply]
Indeed. I think the patch add the VESA complaint mode switching. I also suspect that the mode switching as used by the Trident drivers with higher bit depths could be quickly added to DOSEMU2, but those higher resolution modes are simply not interpreted fully at this point.

**Comment by andrewbird** [comment]
I don't know if you guys follow os2museum but Michal has some interesting articles on building antique Windows display drivers using the SDK for higher screen resolutions.  See https://www.os2museum.com/wp/antique-display-driving/. In the past he's been excellent on releasing (even unfinished) source code if people are interested.

**Reply by jschwartzenberg** [reply]
Yeah I saw his driver. It seems he is also struggling to go beyond 8-bit. I think it would be very interesting to look into using those drivers with DOSEMU2 especially if he manages to overcome that.

**Reply by stsp** [reply]
Yeah, its completely cool that
other guys started to write win31
drivers for us. :)
